File: leetcode/0201-bitwise-and-of-numbers-range/0201-bitwise-and-of-numbers-range.ts
/**
 * Finds the bitwise AND of all numbers in the range [left, right] inclusive.
 * 
 * The algorithm works by finding the common prefix of left and right in their binary representation.
 * It repeatedly clears the rightmost set bit of 'right' until right becomes less than or equal to left.
 * This effectively finds the longest common prefix of all numbers in the range.
 * 
 * @param left - The starting number of the range (inclusive)
 * @param right - The ending number of the range (inclusive)
 * @returns The bitwise AND of all numbers in the range
 */
function rangeBitwiseAnd(left: number, right: number): number {
    // Keep clearing the rightmost set bit of 'right' until it becomes less than or equal to 'left'
    // This finds the common binary prefix of all numbers in the range
    while (left < right) {
        // Clear the rightmost set bit using Brian Kernighan's algorithm
        // right & (right - 1) removes the lowest set bit
        right &= right - 1;
    }
  
    // The remaining value is the bitwise AND of all numbers in the range
    return right;
}

File: 0201-bitwise-and-of-numbers-range/0201-bitwise-and-of-numbers-range.java
class Solution {
    public int rangeBitwiseAnd(int left, int right) {
        // The idea is to find the common prefix of all numbers in range [left, right]
        // We achieve this by repeatedly turning off the rightmost set bit of 'right'
        // until 'right' becomes less than or equal to 'left'
      
        // Continue while right is greater than left
        while (left < right) {
            // Turn off the rightmost set bit in 'right'
            // This operation: right & (right - 1) removes the rightmost 1-bit
            // Example: 12 (1100) & 11 (1011) = 8 (1000)
            right &= (right - 1);
        }
      
        // At this point, 'right' contains the common prefix of all numbers
        // in the range, which is the bitwise AND result
        return right;
    }
}
